### PR TITLE
Issue #13809: Kill mutation ConfigurationLoader2

### PR DESCRIPTION
**Pull request by Kevin222004**
Issue #13809: Kill mutation ConfigurationLoader2

Test

### CODE REVIEW

File: src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java [hunk 1]
```diff
@@ -769,4 +769,20 @@ public void testLoadConfiguration3() throws Exception {
                     .contains("query");
         }
     }
+
+    @Test
+    public void testLengthInReplaceProperties() throws Exception {
+        final String[][] testValues = {
+                {"${a}"},
+                {"x${a}"},
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java:777]: Why multi dimensional array ?

File: src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java [hunk 2]
```diff
@@ -769,4 +769,20 @@ public void testLoadConfiguration3() throws Exception {
...
+        };
+        final Properties props = new Properties();
+        for (String[] testValue : testValues) {
+            final String value = (String) getReplacePropertiesMethod().invoke(
+                    null, testValue[0], new PropertiesExpander(props), "Default");
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java:782]: Can we load whole xml file and check values of Checker property?

File: src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java [hunk 3]
```diff
@@ -769,4 +769,20 @@ public void testLoadConfiguration3() throws Exception {
...
+            assertWithMessage("\"" + testValue[0] + "\"")
+                    .that(value)
+                    .isEqualTo("Default");
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java:785]: Why `{"x${a}"}` becomes Default? 
Should  be `Default`.

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java (hunk 1 above)
Comment (Kevin222004) [reply src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java]: converted into single dimension

File: src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java (hunk 3 above)
Comment (Kevin222004) [reply src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java]: The default value is provided as "Default" thats why

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java]: Should  be `xDefault`.
As x is before variable.

**Comment by romani** [comment]
Kevin requested help on my comments , we need to help him.

**Comment by romani** [comment]
your code is correct by it is really hard to understand why this works like this.

lets make test full functional.
```
<?xml version="1.0" encoding="UTF-8"?>

<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="MemberName">
            <property name="severity" value="${severity}"  default="ignore"/>
        </module>
    </module>
</module>
 
```

load this config as it is done at https://github.com/checkstyle/checkstyle/blob/bd3a9bb9ec9cf804f628171c2e7557b0afbcfd93/src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java#L81
and confirm that property is set to default value.

File: src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java (hunk 3 above)
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java]: I recalling this as default is for whole property value

File: src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java (hunk 1 above)
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java]: need to be converted to more full functional test based on xml files

File: src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java (hunk 2 above)
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java]: same as above, should be like https://github.com/checkstyle/checkstyle/blob/bd3a9bb9ec9cf804f628171c2e7557b0afbcfd93/src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java#L81

**Review by romani — APPROVED** [review]
I updated code to required state

**Comment by romani** [comment]
Github, generate website

File: src/xdocs/config.xml [hunk 4]
```diff
@@ -244,6 +244,11 @@
 &lt;/module&gt;
       </source>
 
+      <p>
+        ATTENTION: default value is applied to module property value if at least one expansion
+        property is not defined.
```
Comment (rnveach) [inline src/xdocs/config.xml:249]: I assume this is referring to property values with multiple expansions in a single property?

This needs to be expanded. In addition why don't we add an example?

Comment (romani) [reply src/xdocs/config.xml:249]: Not only multiple expansion, any .

File: src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java [hunk 5]
```diff
@@ -769,4 +769,24 @@ public void testLoadConfiguration3() throws Exception {
                     .contains("query");
         }
     }
+
+    @Test
+    public void testDefaultValuesForNonDefinedProperties() throws Exception {
```
[the remark this answers is not in the document]
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/ConfigurationLoaderTest.java:774]: Whole test class is not using our general format. I still do not feel good to migrate this also to our general for modules format.

Tests are very similar to each other.

File: src/xdocs/config.xml (hunk 4 above)
Comment (romani) [reply src/xdocs/config.xml:249]: I thought about example, but it is not easy to explain and might be too technical, and looks like nobody dealing with this, so we found this nuance only during pitest.

I think we don't need to make our doc to be too much extended and hard to read.

**Pull request merged**